File: platform/fastapi_server/routes/healthz_s3_data.route.py
from aws_s3_client import config_from_env
from botocore.config import Config as BotoConfig
from fastapi import FastAPI, Query, Request
from fastapi.responses import JSONResponse


def _build_boto_config(config: dict) -> BotoConfig:
    """Build a botocore Config from the S3 config dict."""
    kwargs: dict = {
        "connect_timeout": config.get("connect_timeout", 10),
        "read_timeout": config.get("read_timeout", 60),
        "retries": {"max_attempts": config.get("max_retries", 3)},
    }
    proxy_url = config.get("proxy_url")
    if proxy_url:
        kwargs["proxies"] = {"http": proxy_url, "https": proxy_url}
    return BotoConfig(**kwargs)


def _get_yaml_s3(app) -> dict | None:
    """Get storage.s3 section from AppYamlConfig, or None if unavailable."""
    app_config = getattr(getattr(app, "state", None), "config", None)
    if app_config and hasattr(app_config, "get_nested"):
        return app_config.get_nested("storage", "s3")
    return None


def mount(app: FastAPI):
    """
    Mount S3 data-exploration routes.
    Read-only endpoints for browsing buckets, objects, and object metadata.
    """

    async def _create_client(request_app=None):
        """Create a short-lived S3 client from environment config."""
        from aiobotocore.session import get_session

        yaml_s3 = _get_yaml_s3(request_app or app)
        config = config_from_env(yaml_config=yaml_s3)
        session = get_session()
        client_kwargs: dict = {
            "region_name": config["region"],
            "config": _build_boto_config(config),
            "verify": config.get("verify_ssl", False),
        }
        if config["endpoint_url"]:
            client_kwargs["endpoint_url"] = config["endpoint_url"]
        if config["aws_access_key_id"] and config["aws_secret_access_key"]:
            client_kwargs["aws_access_key_id"] = config["aws_access_key_id"]
            client_kwargs["aws_secret_access_key"] = config["aws_secret_access_key"]

        client = await session.create_client("s3", **client_kwargs).__aenter__()
        return client, config

    @app.get("/healthz/s3/buckets")
    async def s3_buckets(request: Request):
        client = None
        try:
            client, config = await _create_client(request.app)
            response = await client.list_buckets()

            buckets = [
                {
                    "name": b.get("Name"),
                    "creationDate": b["CreationDate"].isoformat()
                    if b.get("CreationDate")
                    else None,
                }
                for b in response.get("Buckets", [])
            ]

            return {
                "configuredBucket": config["bucket_name"] or None,
                "buckets": buckets,
            }
        except Exception as e:
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )
        finally:
            if client:
                await client.__aexit__(None, None, None)

    @app.get("/healthz/s3/buckets/{bucket}/objects")
    async def s3_bucket_objects(
        bucket: str,
        request: Request,
        prefix: str = "",
        cursor: str = "",
        maxKeys: int = 50,
    ):
        max_keys_val = min(200, max(1, maxKeys))
        client = None
        try:
            client, _config = await _create_client(request.app)

            params = {"Bucket": bucket, "MaxKeys": max_keys_val}
            if prefix:
                params["Prefix"] = prefix
            if cursor:
                params["ContinuationToken"] = cursor

            response = await client.list_objects_v2(**params)

            objects = [
                {
                    "key": obj.get("Key"),
                    "size": obj.get("Size"),
                    "lastModified": obj["LastModified"].isoformat()
                    if obj.get("LastModified")
                    else None,
                    "storageClass": obj.get("StorageClass"),
                }
                for obj in response.get("Contents", [])
            ]

            common_prefixes = [
                p.get("Prefix") for p in response.get("CommonPrefixes", [])
            ]

            return {
                "bucket": bucket,
                "prefix": prefix or None,
                "nextCursor": response.get("NextContinuationToken"),
                "isTruncated": response.get("IsTruncated", False),
                "keyCount": response.get("KeyCount", 0),
                "objects": objects,
                "commonPrefixes": common_prefixes,
            }
        except Exception as e:
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )
        finally:
            if client:
                await client.__aexit__(None, None, None)

    @app.get("/healthz/s3/buckets/{bucket}/objects/metadata")
    async def s3_object_metadata(
        bucket: str,
        request: Request,
        key: str = Query(..., description="S3 object key"),
    ):
        object_key = key
        if not object_key:
            return JSONResponse(
                status_code=400,
                content={"error": "Object key is required"},
            )

        client = None
        try:
            client, _config = await _create_client(request.app)
            response = await client.head_object(Bucket=bucket, Key=object_key)

            return {
                "bucket": bucket,
                "key": object_key,
                "contentType": response.get("ContentType"),
                "contentLength": response.get("ContentLength", 0),
                "lastModified": response["LastModified"].isoformat()
                if response.get("LastModified")
                else None,
                "eTag": response.get("ETag"),
                "storageClass": response.get("StorageClass"),
                "metadata": response.get("Metadata", {}),
                "serverSideEncryption": response.get("ServerSideEncryption"),
            }
        except Exception as e:
            error_str = str(e).lower()
            if "404" in error_str or "not found" in error_str or "nosuchkey" in error_str:
                return JSONResponse(
                    status_code=404,
                    content={"error": f"Object not found: {object_key}"},
                )
            return JSONResponse(
                status_code=500,
                content={"error": str(e)},
            )
        finally:
            if client:
                await client.__aexit__(None, None, None)

File: platform/fastify_server/routes/healthz_s3_data.route.mjs
import {
  S3Client,
  ListBucketsCommand,
  ListObjectsV2Command,
  HeadObjectCommand,
} from "@aws-sdk/client-s3";
import { NodeHttpHandler } from "@smithy/node-http-handler";
import { configFromEnv } from "aws-s3-client";

/**
 * Mount S3 data-exploration routes.
 * Read-only endpoints for browsing buckets, objects, and object metadata.
 * @param {import('fastify').FastifyInstance} server
 */
export async function mount(server) {
  // Helper: create a short-lived S3 client, destroy when done
  async function createClient() {
    const yamlS3 = server.config?.getNested?.(['storage', 's3']) ?? null;
    const config = configFromEnv(undefined, yamlS3);
    const clientConfig = { region: config.region };
    if (config.endpointUrl) {
      clientConfig.endpoint = config.endpointUrl;
    }
    clientConfig.forcePathStyle = config.forcePathStyle ?? !!config.endpointUrl;
    if (config.awsAccessKeyId && config.awsSecretAccessKey) {
      clientConfig.credentials = {
        accessKeyId: config.awsAccessKeyId,
        secretAccessKey: config.awsSecretAccessKey,
      };
    }
    // Proxy support via HttpOptions.agent
    if (config.proxyUrl) {
      const { HttpsProxyAgent } = await import("https-proxy-agent");
      const agent = new HttpsProxyAgent(config.proxyUrl);
      clientConfig.requestHandler = new NodeHttpHandler({
        httpAgent: agent,
        httpsAgent: agent,
      });
    }
    return { client: new S3Client(clientConfig), config };
  }

  async function withClient(fn) {
    const { client, config } = await createClient();
    try {
      return await fn(client, config);
    } finally {
      try {
        client.destroy();
      } catch (_) {
        // ignore
      }
    }
  }

  // ── GET /healthz/s3/buckets ──
  server.get("/healthz/s3/buckets", async (_request, reply) => {
    try {
      return await withClient(async (client, config) => {
        const response = await client.send(new ListBucketsCommand({}));

        const buckets = (response.Buckets || []).map((b) => ({
          name: b.Name,
          creationDate: b.CreationDate?.toISOString() || null,
        }));

        return {
          configuredBucket: config.bucketName || null,
          buckets,
        };
      });
    } catch (err) {
      reply.code(500);
      return { error: err.message || String(err) };
    }
  });

  // ── GET /healthz/s3/buckets/:bucket/objects?prefix=&cursor=&maxKeys=50 ──
  server.get("/healthz/s3/buckets/:bucket/objects", async (request, reply) => {
    const { bucket } = request.params;
    const prefix = request.query.prefix || "";
    const cursor = request.query.cursor || undefined;
    const maxKeys = Math.min(200, Math.max(1, parseInt(request.query.maxKeys, 10) || 50));

    try {
      return await withClient(async (client) => {
        const params = {
          Bucket: bucket,
          MaxKeys: maxKeys,
        };
        if (prefix) params.Prefix = prefix;
        if (cursor) params.ContinuationToken = cursor;

        const response = await client.send(new ListObjectsV2Command(params));

        const objects = (response.Contents || []).map((obj) => ({
          key: obj.Key,
          size: obj.Size,
          lastModified: obj.LastModified?.toISOString() || null,
          storageClass: obj.StorageClass || null,
        }));

        // Derive common prefixes (virtual folders)
        const commonPrefixes = (response.CommonPrefixes || []).map(
          (p) => p.Prefix
        );

        return {
          bucket,
          prefix: prefix || null,
          nextCursor: response.NextContinuationToken || null,
          isTruncated: response.IsTruncated || false,
          keyCount: response.KeyCount || 0,
          objects,
          commonPrefixes,
        };
      });
    } catch (err) {
      reply.code(500);
      return { error: err.message || String(err) };
    }
  });

  // ── GET /healthz/s3/buckets/:bucket/objects/metadata?key=<objectKey> ──
  // Object key passed as query param since S3 keys can contain slashes
  server.get("/healthz/s3/buckets/:bucket/objects/metadata", async (request, reply) => {
    const { bucket } = request.params;
    const objectKey = request.query.key;

    if (!objectKey) {
      reply.code(400);
      return { error: "Object key is required" };
    }

    try {
      return await withClient(async (client) => {
        const response = await client.send(
          new HeadObjectCommand({ Bucket: bucket, Key: objectKey })
        );

        return {
          bucket,
          key: objectKey,
          contentType: response.ContentType || null,
          contentLength: response.ContentLength || 0,
          lastModified: response.LastModified?.toISOString() || null,
          eTag: response.ETag || null,
          storageClass: response.StorageClass || null,
          metadata: response.Metadata || {},
          serverSideEncryption: response.ServerSideEncryption || null,
        };
      });
    } catch (err) {
      const errStr = String(err).toLowerCase();
      if (errStr.includes("404") || errStr.includes("not found") || errStr.includes("nosuchkey")) {
        reply.code(404);
        return { error: `Object not found: ${objectKey}` };
      }
      reply.code(500);
      return { error: err.message || String(err) };
    }
  });
}
